File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        HashMap<Character,Character> ms = new HashMap<>();
        ms.put('(',')');
        ms.put('{','}');
        ms.put('[',']');
        Stack<Character> charStack = new Stack<>();
        for (char key : s.toCharArray()) {
            if (ms.containsKey(key)) {
                charStack.push(key);
            } else {
                if (charStack.isEmpty()) {
                    return false;
                }
                char popped = charStack.pop();
                if (!(ms.get(popped) == key)) {
                    return false;
                }
            }
        }
        return charStack.isEmpty() ? true : false;
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
   stack = []
    brackets = { ')' => '(', ']' => '[', '}' => '{' }
    s.each_char do |char|
      if brackets.values.include?(char)
        stack.push(char)
      elsif brackets.keys.include?(char)
        return false if stack.empty? || stack.pop != brackets[char]
      else
        return false
      end
    end
    return stack.empty? 
end